components/collections/src/lib.rs

```diff
@@ -1,15 +1,315 @@
 // Copyright 2017 TiKV Project Authors. Licensed under Apache-2.0.
 
+#![feature(hash_drain_filter)]
+
 #[allow(unused_extern_crates)]
 extern crate tikv_alloc;
 
-use std::{cmp::Eq, hash::Hash};
+use std::{
+    borrow::Borrow,
+    cmp::Eq,
+    collections,
+    fmt::{self, Debug},
+    hash::{self, Hash},
+    ops::Index,
+};
 
-pub type HashMap<K, V> =
-    std::collections::HashMap<K, V, std::hash::BuildHasherDefault<fxhash::FxHasher>>;
-pub type HashSet<T> = std::collections::HashSet<T, std::hash::BuildHasherDefault<fxhash::FxHasher>>;
-pub use std::collections::hash_map::Entry as HashMapEntry;
+type HashBuilder = hash::BuildHasherDefault<fxhash::FxHasher>;
+pub struct HashMap<K, V> {
+    base: collections::HashMap<K, V, HashBuilder>,
+}
+pub type HashSet<T> = collections::HashSet<T, HashBuilder>;
+pub use collections::hash_map::Entry as HashMapEntry;
 
 pub fn hash_set_with_capacity<T: Hash + Eq>(capacity: usize) -> HashSet<T> {
     HashSet::with_capacity_and_hasher(capacity, fxhash::FxBuildHasher::default())
 }
+
+impl<K, Q: ?Sized, V> Index<&Q> for HashMap<K, V>
+where
+    K: Eq + Hash + Borrow<Q>,
+    Q: Eq + Hash + Debug,
+{
+    type Output = V;
+
+    fn index(&self, key: &Q) -> &V {
+        self.base
+            .get(key)
+            .unwrap_or_else(|| panic!("no entry found for key: {:?}", key))
```

Comment:
I'm not fully sure if it worths the effort to wrap a new HashMap just for printing a key, which itself may not be useful in production if the key has customer's sensitive information, as we would require them be redacted. And the wrap code would need to be maintained if HashMap has new APIs introduced. 

Maybe just still use the original HashMap and print it in caller side when it needs to be panic with proper redact log called.

Reply:
+1, maybe forbid using index method forcing caller to use `get`

Reply:
In fact, I am planning to add auto-shrink on remove to address OOM issues like #15458

Reply:
TiKV handles redact on `impl Debug`, this panic will not leak sensitive info.

Reply:
+1, it's unnecessary to import a new HashMap just used for printing the Key.